modular_zubbers/code/modules/holomaps/holomap.dm

```diff
@@ -0,0 +1,320 @@
+#define HOLOMAP_LOW_LIGHT 1, 2
+#define HOLOMAP_HIGH_LIGHT 2, 3
+#define HOLOMAP_LIGHT_OFF 0
+
+// Wall mounted holomap of the station
+// Credit to polaris for the code which this current map was originally based off of, and credit to VG for making it in the first place.
+
+/obj/machinery/holomap
+	name = "\improper holomap"
+	desc = "A virtual map of the surrounding area."
+	icon = 'modular_zubbers/icons/obj/machines/holomap/stationmap.dmi'
+	icon_state = "station_map"
+	layer = ABOVE_WINDOW_LAYER
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 10
+	active_power_usage = 100
+	light_color = HOLOMAP_HOLOFIER
+
+	/// The mob beholding this marvel.
+	var/mob/watching_mob
+	/// The image that can be seen in-world.
+	var/image/small_station_map
+	/// The little "map" floor painting.
+	var/image/floor_markings
+
+	// zLevel which the map is a map for.
+	var/current_z_level
+	/// The various images and icons for the map are stored in here, as well as the actual big map itself.
+	var/datum/station_holomap/holomap_datum
+
+	var/wall_frame_type = /obj/item/wallframe/holomap
+
+	var/construction_state
+
+/obj/machinery/holomap/open
+	panel_open = TRUE
+	construction_state = "unwired"
+	icon_state = "station_map_unwired"
+
+/obj/machinery/holomap/Initialize(mapload)
+	. = ..()
+	current_z_level = z
+	SSholomaps.station_holomaps += src
+
+/obj/machinery/holomap/post_machine_initialize()
+	. = ..()
+	setup_holomap()
+
+/obj/machinery/holomap/Destroy()
+	SSholomaps.station_holomaps -= src
+	close_map()
+	QDEL_NULL(holomap_datum)
+	. = ..()
+
+/obj/machinery/holomap/proc/setup_holomap()
+	var/turf/current_turf = get_turf(src)
+	holomap_datum = new
+	floor_markings = image('modular_zubbers/icons/obj/machines/holomap/stationmap.dmi', "decal_station_map")
+
+	if(!("[HOLOMAP_EXTRA_STATIONMAP]_[current_z_level]" in SSholomaps.extra_holomaps))
+		holomap_datum.initialize_holomap_bogus()
+		update_icon()
+		return
+
+	holomap_datum.bogus = FALSE
+	holomap_datum.initialize_holomap(current_turf.x, current_turf.y, current_z_level, reinit_base_map = TRUE, extra_overlays = handle_overlays())
+
+	update_icon()
+
+/obj/machinery/holomap/attack_hand(mob/user)
+	if(user && user == holomap_datum?.watching_mob)
+		close_map(src)
+		return
+
+	open_map(user, src)
+
+/// Tries to open the map for the given mob. Returns FALSE if it doesn't meet the criteria, TRUE if the map successfully opened with no runtimes.
+/obj/machinery/holomap/proc/open_map(mob/user)
+	if((machine_stat & (NOPOWER | BROKEN)) || !user?.client || panel_open || user.hud_used.holomap.used_station_map)
+		return FALSE
+
+	if(!holomap_datum)
+		// Something is very wrong if we have to un-fuck ourselves here.
+		message_admins("\[HOLOMAP] WARNING: Holomap at [x], [y], [z] [ADMIN_FLW(src)] had to set itself up on interact! Something during Initialize went very wrong!")
```

Comment:
message_admins doesn't create a log entry, and there's no guarantee any admin is going to see this, nor care, nor know what to do.